File: clojure/4.clj
; A palindromic number reads the same both ways. The largest palindrome made
; from the product of two 2-digit numbers is 9009 = 91 × 99.
;
; Find the largest palindrome made from the product of two 3-digit numbers.
;
; Answer: 906609

(defn palindrom? [s]
  (let [size (count s)
        middle (int (/ size 2))
        pre (.substring s 0 middle)
        i (if (even? size) middle (inc middle))]
    (= (list* pre) (reverse (.substring s i size)))))


(defn palindrom-numbers []
  (for [a (range 100 999)
        b (range 100 999)
        :when (palindrom? (format "%d" (* a b)))]
    (* a b)))

(prn (apply max (palindrom-numbers)))

File: go/4.go
/*
A palindromic number reads the same both ways. The largest palindrome made from
the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

Answer: 906609
*/
package main

import "fmt"

func isPalindrom(n int) bool {
	str := fmt.Sprintf("%d", n)
	size := len(str)
	for i := 0; i < size/2; i++ {
		if str[i] != str[size-i-1] {
			return false
		}
	}
	return true
}

func e4() {
	max := 0
	for i := 1; i < 1000; i++ {
		for j := 1; j < 1000; j++ {
			n := i * j
			if isPalindrom(n) && n > max {
				max = n
			}
		}
	}

	fmt.Println(max)
}

func init() {
    register("4", e4)
}
